### DISCUSSION TITLE
WebSocket / Socket.IO implementation.

### DISCUSSION THREAD

**Discussion by omershevach** [vikejs/vike#906] [Discussion]
I would really like to know how to implement WebSocket / Socket.IO in vite plugin ssr.
i working on a project that require real time updates, could be that sockets is not the best solution or maybe its better to create the app in SPA/Next.js which could be more suitable but let me know.

ive tried using client routing, which seems like the right direction but how can you use a websocket in a way that will persist on all routes and will not get disconnect and connected on each navigation?

**Comment by brillout** [comment]
Use the `onHydrationEnd()` hook and/or a global client-side variable e.g. `window.initSocket` that you init exactly once.

**Reply by omershevach** [reply]
from what i know onHydrationEnd hook gets called on each navigation.

**Reply by brillout** [reply]
No it shouldn't. If it does then something's wrong with your client routing setup.

**Reply by omershevach** [reply]
![image](https://github.com/brillout/vite-plugin-ssr/assets/55285791/bf69f411-f4f1-4640-9732-28d5565ad398)
navigating trough this routes (with preserve log):
"/" -> "/test" -> "/"

**Reply by omershevach** [reply]
so this isn't normal?

**Reply by brillout** [reply]
It isn't. See `onHydrationEnd()` at `/examples/react-full/`.

**Reply by omershevach** [reply]
i see now, thanks.
by the way, would you recomend VPS for large scale project?

**Reply by brillout** [reply]
Yes, it's already used by large-scale projects.

FYI you can [sponsor](https://github.com/sponsors/brillout) and get high priority support.

**Reply by opudalo** [reply]
@GetNek if you need to preserve log between navigations (to see the logs), that means that you were doing a full page reload, and NOT client-side navigation. Also that is the reason why `onHydrationEnd` fires more than once (each time for full page load).
<img width="191" alt="image" src="https://github.com/brillout/vite-plugin-ssr/assets/775750/d18f6212-6164-46e5-b164-7fb840b83fb3">

also also, 
via global variable:
```js
// page that needs socket 
const Page = () => {
   // take socket from global var or initialize it and store it in global variable for future navigations 
   const socketManager = window._smCache =  window._smCache || new SocketManager()
   return <App socket={socketManager} />
}

```
you can use global variable and store your socket connection from whatever there

**Comment by brillout** [comment]
Example by @phiberber: https://github.com/phiberber/vps-react-socket-io-v1.
